src/execute.cxx

```diff
@@ -171,10 +171,35 @@ int ExecuteCommand::PipeChildToStdErr()
     for (;;)
     {
         bSuccess = ReadFile( this->ChildStdErr_Rd, chBuf, BUFSIZE, &dwRead, NULL);
-        if( ! bSuccess || dwRead == 0 ) break;
-
-        bSuccess = WriteFile(hParentOut, chBuf,
-                            dwRead, &dwWritten, NULL);
+        // For an explanation behind the use of termianted here
+        // see the docstring on pipechildtostdout
+        if( ! bSuccess || (dwRead == 0 && this->terminated) ) break;
+        if(dwRead != 0) {
+            bSuccess = WriteFile(hParentOut, chBuf,
+                                dwRead, &dwWritten, NULL);
+            if (dwWritten < dwRead && bSuccess){
+                // incomplete write but not a failure
+                // since bSuccess is true
+                // So lets write until bSuccess is false or
+                // until all bytes are written
+                int currentPos = dwWritten;
+                while((dwWritten < dwRead) || dwWritten == 0) {
+                    dwRead = dwRead - dwWritten;
+                    CHAR * partialBuf = new CHAR[dwRead];
+                    for(int i = 0; i < dwRead; ++i) {
+                        partialBuf[i] = chBuf[currentPos + i];
+                    }
+                    bSuccess = WriteFile(hParentOut, partialBuf,
+                                        dwRead, &dwWritten, NULL);
+                    delete partialBuf;
+                    if (! bSuccess) break;
+                    currentPos += dwWritten;
+                }
+            }
+            if (! bSuccess ) {
+                break;
+            }
+        }
```

Comment:
Would it be possible to consolidate this with the logic in `PipeChildToStdout`? I think one function could read a file descriptor in a loop (and each of these could pass in the file descriptor of interest to that function).

Reply:
Syncing PR with out of band discussion: This will be handled as a part of a larger refactor in a follow up PR